### PR TITLE
Add option to add an initial guess for the SLEPc eigensolver

### PR DESCRIPTION
**Pull request by juan42**
Pull request for issue #1130 

I decided to just implement the EigenSolver interface for now and just used eigenproblems_ex2 for the demonstration to keep eigenproblems_ex1 as simple as possible.

EDIT: As i'm thinking about it afterwards, a new example with two solves and checking for the number of iterations would make sene.

### CODE REVIEW

**Comment by jwpeterson** [comment]
`EPSSetInitialVector` has been in SLEPc since v3.1, aka 2010 time frame:

```
commit 5f7b509317e2c0b9c87c93529abfe48dd64f4899
Author: Jose E. Roman <jroman@dsic.upv.es>
Date:   Thu Apr 1 13:46:42 2010 +0000

    Replaced EPSSetInitialVector by the new function EPSSetInitialSpace; still more work required for left initial vectors
    
    Subversion-commit: r1933
```

So we'll need to guard this function call, i.e.
```
#if SLEPC_VERSION_LESS_THAN(3,1,0)
  libmesh_error_msg("SLEPc 3.1 is required to call EigenSolver::set_initial_space() ");
#else
// current PR code.
#endif
```

File: src/solvers/slepc_eigen_solver.C [hunk 1]
```diff
@@ -743,6 +743,18 @@ void SlepcEigenSolver<T>::attach_deflation_space(NumericVector<T> & deflation_ve
 }
 
 template <typename T>
+void SlepcEigenSolver<T>::set_initial_space(NumericVector<T> & initial_space_in)
+{
+  this->init();
+  
+  PetscErrorCode ierr = 0;
+  Vec initial_vector = (cast_ptr<PetscVector<T> *>(&initial_space_in))->vec();
```
Comment (jwpeterson) [inline src/solvers/slepc_eigen_solver.C]: (Just a reminder to self, this is not something for you to necessarily fix.)

I see that this idiom is used in several places in the `SlepcEigenSolver` code, but in optimized mode, `cast_ptr` simply does a `static_cast`, which could cause this line of code to segfault somewhat mysteriously if a non-PETSc `NumericVector` is passed in.  I'd like to change these to be done in two lines with a dynamic cast, checking the result.

**Comment by jwpeterson** [comment]
Thanks for making the changes.  I think I will squash these commits together before merging.  Then @BalticPinguin can perhaps rebase his related changes on master and try his PR again...

**Comment by BalticPinguin** [comment]
So I removed the initiial_guess related part. No rebase needed from my side.

**Comment by jwpeterson** [comment]
Merged to master in b64cb69.  I also fixed some whitespace issues that were introduced by the patches during merge...

**Pull request closed without merging**